[GeoMechanicsApplication] gtests

markelov208 opened the pull request:
**📝 Description**
A brief description of the PR.

Please mark the PR with appropriate tags: 
- The PR has been created to check the differences

**🆕 Changelog**
Please summarize the changes in one list to generate the changelog:
E.g.
- Added feature X to Y
- Added Foo Application
- Fixed X (#XXXX Reference to issue if apply) 
- etc...

roigcarlo committed:
Prepare to build gtests

roigcarlo committed:
Removing references to runkratos

roigcarlo committed:
Enable gtesting

roigcarlo committed:
Add target for the core

roigcarlo committed:
Moving mpi related test utilities out of the core

roigcarlo committed:
Fixing mpi dependant tests

roigcarlo committed:
Refactor core and mpicore tests

roigcarlo committed:
Updating tests

roigcarlo committed:
Adding gtest

roigcarlo committed:
Patch for cmake 3.27

roigcarlo committed:
Mpi tests

roigcarlo committed:
Adding kernel destructor and fixing mpi test with individual kernel instances

roigcarlo committed:
Restoring internal macros checks

roigcarlo committed:
Fixing problem comparing array_1d with ZeroVector(x) from boost

roigcarlo committed:
Fixing cstr cmp in text excpetion

roigcarlo committed:
Fixing cstr cmp in code location

roigcarlo committed:
Test_checks now checks the checks class. Checkmate

roigcarlo committed:
Fixing logger test (bad copy paste)

roigcarlo committed:
Merge branch 'master' into ci/gtests

roigcarlo committed:
Min cmake version

roigcarlo committed:
Missing include

roigcarlo committed:
Using distributed expects

roigcarlo committed:
Merge branch 'master' into ci/gtests

roigcarlo committed:
Removing testing_ folder

roigcarlo committed:
Merge branch 'master' into ci/gtests

roigcarlo committed:
Merge conflicts

Commit:
declared size_t

Commit:
compilable version without MPI

Commit:
added testing folder

Commit:
Merge branch 'master' into ci/gtests

# Conflicts:
#	CMakeLists.txt
#	kratos/CMakeLists.txt
#	kratos/includes/global_pointer.h
#	kratos/mpi/CMakeLists.txt
#	kratos/mpi/tests/cpp_tests/utilities/test_search_utilities.cpp
#	kratos/testing/testing.h
#	kratos/tests/cpp_tests/processes/test_set_initial_state_process.cpp
#	kratos/tests/test_utilities/test_constitutive_law.cpp
#	scripts/wheels/linux/configure.sh
#	scripts/wheels/linux/configure_mpi.sh
#	scripts/wheels/windows/configure.bat

Commit:
added file(GLOB_RECURSE KRATOS_TEST_SOURCES

Commit:
changed MPISynchronousPointSynchronization

Commit:
made comparison of pointers before calling the macros

Commit:
moved cpp_tests_utilities.* from utilities to tests/cpp_tests/utilities

Commit:
added USE_MPI OFF to test Deltares TC

Commit:
corrected include for test_suite.h

Commit:
added tests/cpp_tests/*.cpp to KRATOS_TEST_UTILITIES_SOURCES to compile tests with MPI

Commit:
Merge branch 'master' into geo/gtests

# Conflicts:
#	CMakeLists.txt
#	applications/PfemSolidMechanicsApplication/custom_problemtype/Kratos_Pfem_Solid_Mechanics_Application.gid/002_Kratos_Pfem_Solid_Mechanics_Application_aux.win.bas
#	docs/pages/Kratos/For_Developers/General/How-to-make-Git-ignore-the-files-resulting-from-a-compilation-without-conflicts-in-.gitignore.md
#	kratos/python_scripts/testing/run_tests.py
#	kratos/run_kratos/CMakeLists.txt

Commit:
added include of test_suite.h

Commit:
- moved cpp_tests_utilities.* to tests/test_utilities

- added gtest to CMakeLists.txt
- added KRATOS_API(GE_MECHANICS_APPLICATION)
- added KRATOS_API(STRUCTURAL_MECHNAICS_APPLICATION)
- added KRATOS_API(KRATOS_CORE)
- removed KRATOS_API(KRATOS_CORE)
- moved scoped_file.* to
- added Kratos*FastSuite classes in testing.h
- moved  test_element.* and test_constitutive_law.* to tests/cpp_tests/utilities
- replaced KRATOS_SKIP_TEST_IF_NOT with KRATOS_ERROR_IF_NOT in test_integration_values_extrapolation_to_nodes_process.cpp

Commit:
commented Factory, it seems it is not used in Kratos yet.

Commit:
modifications in CMakeLists.txt files

- removed running gtest automatically
- added TestBarElement manually to KratosCoreTest.exe

Commit:
CMake files

- changed libs to lib because libs is deleted
- changed KratosStructuralMechanicsTests to KratosStructuralMechanicsTest to be consistent with other executables

roigcarlo commented on CMakeLists.txt:
```diff
@@ -651,8 +651,8 @@ kratos_python_install(${INSTALL_PYTHON_USING_LINKS} "${KRATOS_SOURCE_DIR}/kratos
 
 # Install the libraries in the libs folder
-install(FILES ${Boost_LIBRARIES} DESTINATION libs)
-install(FILES ${EXTRA_INSTALL_LIBS} DESTINATION libs)
```
@markelov208 Please do not change this.

Libs needs to be deleted and repopulated at every compilation to avoid mismatching versions of the same library. This change is intentional.

markelov208 replied:
@roigcarlo , thank you for letting me know. Currently, the libs folder is populated and then deleted. 
[39/40] Install the project...
-- Install configuration: "Debug"
-- Deleting: D:/Checkouts2/KratosProjects/alpha/bin/Debug/KratosMultiphysics
-- Deleting: D:/Checkouts2/KratosProjects/alpha/bin/Debug/libs

Build finished

Please could you let me know how to change the order in cmake file? Thank you.

roigcarlo replied:
The log its a little bit misleading, its the other way around, libs folders gets deleted and then populated if the compilation goes well

markelov208 replied:
Unfortunately, this is exactly what I have on my laptop: files are copied into libs folder, then libs is deleted. ;) I have reverted libs and put REMOVE_INSTALL_DIRECTORIES OFF.

roigcarlo replied:
Mmm that's quite odd. In any case, its important to maintain the name `libs`, so if it fails as last resource we can comment the lines that remove the contents of the folder, but changing the name will effectively break the release and RPATH so that cannot be changed

Commit:
reverted libs

Commit:
modifications

- added geo_testing.h
- added structural_testing.h
- changed const std::vector<double> to Vector
- added KRATOS_ERROR_IN_NOT in test_total_lagrangian_mixed_volumetric_strain_element.cpp
- changed KRATOS_ERROR_IF to EXPECT_THAT in expect.h
- removed KratosStructural* and KratosGeo* from testing.h

Commit:
added KRATOS_SOURCE to working directories

Commit:
removed second definition of KRATOS_TEST_CASE

Commit:
changed comparison of strings

philbucher commented on kratos/tests/cpp_tests/sources/test_checks.cpp:
```diff
@@ -35,8 +35,8 @@ namespace Kratos {
 			KRATOS_EXPECT_NE(1.0, 2.0);
 
-			KRATOS_EXPECT_EQ("Test", "Test");
-			KRATOS_EXPECT_NE("Test ", "Test");
```
this change does not seem right, should work either way. Can you elaborate please?

markelov208 replied:
In this branch KRATOS_EXPECT_EQ is replaced with EXPECT_EQ which tests memory allocations for C strings. I guess these two lines were used to check KRATOS_EXPECT_* functionality for C strings. Perhaps, it is better to remove them now.

philbucher commented on applications/GeoMechanicsApplication/tests/cpp_tests/test_GeoMechanicsNewtonRaphsonErosionProcessStrategy.cpp:
```diff
@@ -26,10 +26,17 @@
+{
+    auto projectDirectory = "./applications/GeoMechanicsApplication/tests/test_compare_sellmeijer/HeightAquiferD10L30.gid";
+    auto kratos_source_directory = std::getenv("KRATOS_SOURCE");
+    return (std::string)kratos_source_directory + "/" + (std::string)projectDirectory;
```
I strongly recommend to use `std::filesystem::path` for anything related to file paths

markelov208 replied:
Thank you for pointing. string is replaced with path.

Commit:
removed USE_MPI set

Commit:
changed size_t to unsigned long to compile on Windows

Commit:
added/removed KRATOS_API(KRATOS_CORE)

Commit:
commented KRATOS_TEST_CASE_IN_SUITE to avoid the macro redefinition warning

Commit:
used std::filesystem::path instead of std::string
